examples/flax/language-modeling/run_bart_dlm_flax.py

```diff
@@ -0,0 +1,955 @@
+        rng, input_rng = jax.random.split(rng)
+
+        # Generate an epoch by shuffling sampling indices from the train dataset
+        num_train_samples = len(tokenized_datasets["train"])
+        # Avoid using jax.numpy here in case of TPU training
+        train_samples_idx = np.random.permutation(np.arange(num_train_samples))
+        train_batch_idx = generate_batch_splits(train_samples_idx, train_batch_size)
+
+        # Gather the indexes for creating the batch and do a training step
+        for step, batch_idx in enumerate(tqdm(train_batch_idx, desc="Training...", position=1)):
+            samples = [tokenized_datasets["train"][int(idx)] for idx in batch_idx]
+            model_inputs = data_collator(samples)
+
+            # Model forward
+            model_inputs = shard(model_inputs.data)
+            state, train_metric, dropout_rngs = p_train_step(state, model_inputs, dropout_rngs)
+            train_metrics.append(train_metric)
+
+            cur_step = epoch * (num_train_samples // train_batch_size) + step
+
+            if cur_step % training_args.logging_steps == 0 and cur_step > 0:
+                # Save metrics
+                train_metric = jax_utils.unreplicate(train_metric)
+                train_time += time.time() - train_start
+                if has_tensorboard and jax.process_index() == 0:
+                    write_train_metric(summary_writer, train_metrics, train_time, cur_step)
+
+                epochs.write(
+                    f"Step... ({cur_step} | Loss: {train_metric['loss']}, Learning Rate:"
+                    f" {train_metric['learning_rate']})"
+                )
+
+                train_metrics = []
+
+            if cur_step % training_args.eval_steps == 0 and cur_step > 0:
+                # ======================== Evaluating ==============================
+                num_eval_samples = len(tokenized_datasets["validation"])
+                # Avoid using jax.numpy here in case of TPU training
+                eval_samples_idx = np.arange(num_eval_samples)
+                eval_batch_idx = generate_batch_splits(eval_samples_idx, eval_batch_size)
+
+                eval_metrics = []
+                for i, batch_idx in enumerate(tqdm(eval_batch_idx, desc="Evaluating ...", position=2)):
+                    samples = [tokenized_datasets["validation"][int(idx)] for idx in batch_idx]
+                    model_inputs = data_collator(samples)
+
+                    # Model forward
+                    model_inputs = shard(model_inputs.data)
+                    metrics = p_eval_step(state.params, model_inputs)
+                    eval_metrics.append(metrics)
+
+                # normalize eval metrics
+                eval_metrics = get_metrics(eval_metrics)
+                eval_metrics = jax.tree_map(jnp.sum, eval_metrics)
+                eval_normalizer = eval_metrics.pop("normalizer")
+                eval_metrics = jax.tree_map(lambda x: x / eval_normalizer, eval_metrics)
```

Comment:
Nice! This is a clean way of normalising the `eval_metrics`! 

Technically speaking, in the `train_step`, the `pmap` won't compute a 'true' mean over devices. Here, what you're doing is computing a normalised loss on each device, and then averaging these losses over devices. This isn't strictly equal to summing the losses over all devices, and then dividing by the number of samples. 

Let $K$ denote the number of devices. Denote the loss on the $i$-th device as $L_i$ (`loss.sum()`) and the number of samples $N_i$ (`label_mask.sum()`). In the `loss_fn`, we compute the normalised loss on each device (`loss.sum() / label_mask.sum()`):

$$\bar{L}_i = \frac{L_i}{N_i}$$

and then average over devices with the `pmap`:

$$\mathcal{L} = \frac{1}{K} \sum_{i=1}^{K} \frac{\bar{L}_i}{N_i}$$

Whereas, for a 'true' loss, we should first add up all the losses over devices:

$$L_{tot} = \sum_{i=1}^{K} \bar{L}_i $$

and then divide by the total number of labels:

$$\mathcal{L}' = \frac{L_{tot}}{N} =  \frac{1}{N}\sum_{i=1}^{K} \bar{L}_i $$

where $N$ is the total number of labels:

$$ N = \sum_{i=1}^{K} N_i $$

If we compare the two and ignore the constant $K$ in the `pmap` average:

$$\mathcal{L} =  \sum_{i=1}^{K} \frac{\bar{L}_i}{N_i}$$

$$ \mathcal{L}' =  \frac{1}{N}\sum_{i=1}^{K} \bar{L}_i $$

we see that the losses are in-fact different. The first expression is what you get if you average the losses on each device, then average these terms over devices with a `pmap`. The second expression is a 'true' loss, what you get by summing the losses on each device, summing the losses over devices, and dividing the summed losses by the sum of the labels mask over devices.

Reply:
You're right, we should not use `pmean` in case of masked labels. I actually borrowed this implementation from `run_mlm_flax.py`.

Reply:
Indeed, we should update the other Flax training examples to reflect this!

Reply:
@sanchit-gandhi Sure, this trick should be applied to examples that do not have the same $N_i$ for every batch, including `run_mlm_flax`, `run_bart_dlm_flax`, `run_summarization_flax` and `run_image_captioning_flax`. 
However, the changes should be made in `run_summarization_flax` and `run_image_captioning_flax` are nontrivial because of the existance of `compute_loss` function (which should return only 1 scalar - the loss.sum()). Also, I think we should apply this trick to train_loss to make it comparable to eval_loss. 
I think it's better to do these in another PR post!

Reply:
Great point! When the batch size is fixed (such as in `run_summarization_flax` and `run_image_captioning_flax`) there is no padding, so we are not obliged to normalise the loss by the number of tokens (as this stays fixed for every batch).

We can either use the trick you used in `eval_loss`, or compute a `psum` over devices for the losses and number of tokens and then normalise accordingly: https://github.com/sanchit-gandhi/seq2seq-speech/blob/cfc6d73959486f5bd71c623ddd95843d62f5a614/run_flax_speech_recognition_seq2seq.py#L1252-L1255

Feel free to leave it as is for this PR - the numerical differences end up being small when averaged over many batches and large amounts of data. Would make for a good follow-up PR!

Reply:
Thanks for the nice trick! 

In fact, I have yet to grasp of `batch size is fixed` = `num of tokens stays fixed for every batch`. Could you elaborate on this a little bit?

Reply:
Sorry, ignore that remark! I checked and it applies equally well to both of those! I remember discussing this point with @patil-suraj offline a couple of months ago. IIRC, in the example given for `run_summarization_flax`, the targets are always truncated to `max_target_length`, and so effectively there is no padding. But in general, the `pmean` won't return a 'true' loss.

Note that we can actually bypass the constraint of the `compute_loss` function only returning one scalar (the total loss). If we pass the flag `has_aux=True` to `jax.value_and_grad` we can return a second, auxiliary output (https://jax.readthedocs.io/en/latest/_autosummary/jax.value_and_grad.html?highlight=value_and_grad). If we return two outputs, `loss` and `num_labels` from our `compute_loss` function, we can take gradients wrt the `loss` and also return the `num_labels`:
```python
def compute_loss(params):
    ...
    return loss, num_labels

# take grads wrt to the first output (loss), also return a second output (num_labels)
grad_fn = jax.value_and_grad(compute_loss, has_aux=True)
loss, num_labels, grad = grad_fn(state.params)
```

Reply:
Aha, I unfortunately forgot the `has_aux=True` flag. Thanks for reminding!

Regarding the num of target tokens, what you said makes sense if the target length exceeds `max_target_length`, but what if it is shorter?

Reply:
And do you think PR is ready to merge? @sanchit-gandhi @patrickvonplaten

Reply:
Having provided pointers related to the data collator, @patil-suraj might wish to take a look before merging!